src/java.net.http/share/classes/jdk/internal/net/http/http3/streams/UniStreamPair.java

```diff
@@ -0,0 +1,515 @@
+import java.time.Duration;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+
+import jdk.internal.net.http.Http3Connection;
+import jdk.internal.net.http.common.Logger;
+import jdk.internal.net.http.common.MinimalFuture;
+import jdk.internal.net.http.common.SequentialScheduler;
+import jdk.internal.net.http.common.Utils;
+import jdk.internal.net.http.http3.streams.Http3Streams.StreamType;
+import jdk.internal.net.http.quic.TerminationCause;
+import jdk.internal.net.http.quic.QuicConnection;
+import jdk.internal.net.http.quic.streams.QuicReceiverStream;
+import jdk.internal.net.http.quic.streams.QuicSenderStream;
+import jdk.internal.net.http.quic.streams.QuicStream;
+import jdk.internal.net.http.quic.streams.QuicStreamReader;
+import jdk.internal.net.http.quic.streams.QuicStreamWriter;
+import jdk.internal.net.http.quic.VariableLengthEncoder;
+import static jdk.internal.net.http.http3.Http3Error.H3_STREAM_CREATION_ERROR;
+import static jdk.internal.net.http.quic.TerminationCause.appLayerClose;
+
+/**
+ * A class that models a pair of HTTP/3 unidirectional streams.
+ * This class implements a read loop that calls a {@link
+ * #UniStreamPair(StreamType, QuicConnection, Consumer, Runnable, StreamErrorHandler, Logger)
+ * receiver} every time a {@code ByteBuffer} is read from
+ * the receiver part.
+ * The {@linkplain #futureSenderStreamWriter() sender stream writer},
+ * when available, can be used to write to the sender part.
+ * The {@link #UniStreamPair(StreamType, QuicConnection, Consumer, Runnable,StreamErrorHandler, Logger)
+ * writerLoop} is invoked whenever the writer part becomes unblocked, and
+ * writing can be resumed.
+ * <p>
+ * @apiNote
+ * The creator of the stream pair (typically {@link Http3Connection}) is expected
+ * to complete the {@link #futureReceiverStream()} completable future when the remote
+ * part of the stream pair is created by the remote peer. This class will not
+ * listen directly for creation of new remote streams.
+ * <p>
+ * The {@link QueuingStreamPair} class is a subclass of this class which
+ * implements a writer loop over a unbounded queue of {@code ByteBuffer}, and
+ * can be used when unlimited buffering of data for writing is not an issue.
+ */
+public class UniStreamPair {
+
+    // The sequential scheduler for the local control stream (LCS) writer loop
+    private final SequentialScheduler localWriteScheduler;
+    // The QuicStreamWriter for the local control stream
+    private volatile QuicStreamWriter localWriter;
+    private final CompletableFuture<QuicStreamWriter> streamWriterCF;
+    // a queue of ByteBuffers submitted for writing.
+    // might be null if not used. Only used in QueuingStreamPair.
```

Comment:
It seems to be true. I'll see if I can move it to QueuingStreamPair.

Reply:
Moved in c2a351ed5cdcaa5c0bdf7e28140a1dfb4e751d56 + 91fb530b350f733789c4deae53c83855d00365b6